Add IntField, LongField, FloatField and DoubleField

fcofdez opened the pull request:
This commit adds new IndexableFields that index both points and doc values at once.

Closes #11199

fcofdez committed:
Add IntField, LongField, FloatField and DoubleField

This commit adds new IndexableFields that index both points and doc
values at once.

Closes #11199

jpountz reviewed (commented):
I like it and would be curious to get @rmuir 's thoughts on this PR.

Maybe we could cut over a couple tests that index both points and doc values on the same field to these new `XXXField` classes. I have `TestLongDistanceFeatureQuery`, `TestIndexOrDocValuesQuery` and `TestSortOptimization` in mind, since these tests are about functionality that requires points and doc values to store the same information to work well.

rmuir commented on lucene/core/src/java/org/apache/lucene/document/DoubleField.java:
```diff
@@ -0,0 +1,138 @@
+   * @param sorted configure the field to support multiple DocValues
+   * @throws IllegalArgumentException if the field name or value is null.
+   */
+  public DoubleField(String name, double value, boolean sorted) {
```
Should we just make the field multivalued always to simplify it? Or, make the default `true' and provide another ctor with `boolean multivalued`? I think i'd prefer `multivalued` to `sorted` as the name of the parameter.

gsmiller replied:
+1 to the suggestion of always making it multivalued. What if we always set the docvalue type to `SORTED_NUMERIC`, even if the user is just storing single values? Lucene will optimize the storage if all docs are single-valued, and it's already a common pattern to look for this optimization on read (e.g,. https://github.com/apache/lucene/blob/main/lucene/facet/src/java/org/apache/lucene/facet/LongValueFacetCounts.java#L303).

rmuir commented:
I like the idea too. Make the .document API simpler for typical use-cases! I added a couple cosmetic comments.

fcofdez committed:
Avoid unnecesary boxing and make multivalued fields by default

fcofdez committed:
Add more tests

fcofdez committed:
Add more tests

jpountz commented:
> What if we always set the docvalue type to SORTED_NUMERIC, even if the user is just storing single values? Lucene will optimize the storage if all docs are single-valued, and it's already a common pattern to look for this optimization on read

This sounds good to me. The main issue I see with this is that our default `SortField`s fail if the field is indexed with multi-valued doc values. But maybe we could make it more lenient and sort on the min value for ascending sorts and max values for descending sorts.

fcofdez committed:
Merge remote-tracking branch 'origin/main' into numeric-fields

fcofdez commented:
Thanks all for the reviews! 

> This sounds good to me. The main issue I see with this is that our default SortFields fail if the field is indexed with multi-valued doc values. But maybe we could make it more lenient and sort on the min value for ascending sorts and max values for descending sorts.

I noticed this issue while I was running the sorting tests, maybe we can tackle this in a follow up PR? This one is getting big already, but I'm open to suggestions.

jpountz commented:
Absolutely, adding support for multivalued fields to default sort fields would be a big change too.

gsmiller commented:
> maybe we can tackle this in a follow up PR?

+1

rmuir commented:
There is already `SortedSetSortField` and `SortedNumericSortField` that do this. So I don't understand why it would be a big change? Maybe just add some sugar (e.g. `newSortField` or similar)

jpountz commented:
Good point. I was thinking of making changes to `SortField` but adding factory methods to these Field classes feels like a better approach to me.

rmuir commented:
OK, well maybe defer it if we have options. I haven't thought about the `SortField` approach, maybe it could be more natural that way.

jpountz commented:
I just found an issue with this PR which is that multi-valued floats and doubles are encoded with `Double#doubleToLongBits` while they should use `NumericUtil#doubleToSortableLong` to present ordering within a document.

One way to fix it would be to change the encoding depending on whether the field is multi-valued, but this isn't great as we then also need to pass this flag to factor methods for queries (or sort fields in the future). So maybe we should go with @rmuir's earlier suggestion of only supporting multi-valued fields, so that there is a single encoding in doc values?

jpountz committed:
Merge branch 'main' into numeric-fields

jpountz committed:
Merge remote-tracking branch 'origin/main' into numeric-fields

jpountz committed:
Only support multi-valued fields.

I also updated tests to not duplicate tests across XXXField/XXXPoint but rather
use XXXField classes when it made more sense.

jpountz committed:
Add CHANGES.

jpountz commented:
@fcofdez I hope you don't mind but I checked out your branch to look into the above issue and ended up pushing my changes. These new fields not only support indexing sorted numeric doc values, so that queries do not have to specify whether floats/doubles are indexed via their raw bits or as sortable ints/longs.

fcofdez commented:
@jpountz thanks for looking into this, unfortunately I didn't have time to go through the last round of review comments. The change makes sense to me.

fcofdez reviewed (commented):
@jpountz the latest changes LGTM, thanks for moving this forward 👍

jpountz commented:
Thanks for having a look @fcofdez. I plan on merging this PR soon if there are no objections.

jpountz committed:
Add CHANGES.

jpountz commented:
I cut over the NYC taxis benchmark to these new fields and it yielded a small but consistent speedup. https://home.apache.org/~mikemccand/lucenebench/sparseResults.html#index_throughput
